### PR TITLE
[RISCV] Preserve MMO when expanding PseudoRV32ZdinxSD/PseudoRV32ZdinxLD.

### PR DESCRIPTION
**Pull request by topperc**
This allows the asm printer to print the stack spill/reload messages.

Stacked on #85871

### CODE REVIEW

File: llvm/lib/Target/RISCV/RISCVExpandPseudoInsts.cpp [hunk 1]
```diff
@@ -312,26 +312,40 @@ bool RISCVExpandPseudo::expandRV32ZdinxStore(MachineBasicBlock &MBB,
...
-  BuildMI(MBB, MBBI, DL, TII->get(RISCV::SW))
-      .addReg(Lo, getKillRegState(MBBI->getOperand(0).isKill()))
-      .addReg(MBBI->getOperand(1).getReg())
-      .add(MBBI->getOperand(2));
+  auto MIBLo = BuildMI(MBB, MBBI, DL, TII->get(RISCV::SW))
+                   .addReg(Lo, getKillRegState(MBBI->getOperand(0).isKill()))
+                   .addReg(MBBI->getOperand(1).getReg())
+                   .add(MBBI->getOperand(2));
+
+  MachineMemOperand *MMOHi = nullptr;
```
Comment (wangpc-pp) [inline llvm/lib/Target/RISCV/RISCVExpandPseudoInsts.cpp:320]: Make this a list (can be empty) so we don't need to know if `MMOHi` is nullptr and `setMemRefs` directly?

Comment (topperc) [reply llvm/lib/Target/RISCV/RISCVExpandPseudoInsts.cpp:320]: I guess that would have to be SmallVector so we can push into it?

Comment (wangpc-pp) [reply llvm/lib/Target/RISCV/RISCVExpandPseudoInsts.cpp:320]: Yes! Just a code style suggestion here.

File: llvm/lib/Target/RISCV/RISCVExpandPseudoInsts.cpp [hunk 2]
```diff
@@ -349,36 +359,46 @@ bool RISCVExpandPseudo::expandRV32ZdinxLoad(MachineBasicBlock &MBB,
...
   bool IsOp1EqualToLo = Lo == MBBI->getOperand(1).getReg();
   // Order: Lo, Hi
   if (!IsOp1EqualToLo) {
-    BuildMI(MBB, MBBI, DL, TII->get(RISCV::LW), Lo)
-        .addReg(MBBI->getOperand(1).getReg())
-        .add(MBBI->getOperand(2));
+    auto MIBLo = BuildMI(MBB, MBBI, DL, TII->get(RISCV::LW), Lo)
+                     .addReg(MBBI->getOperand(1).getReg())
+                     .add(MBBI->getOperand(2));
+    MIBLo.setMemRefs(MMOLo);
```
Comment (wangpc-pp) [inline llvm/lib/Target/RISCV/RISCVExpandPseudoInsts.cpp:376]: Remove `MIBLo` and chain the call `setMemRefs` after `BuildMI`?

**Comment by topperc** [comment]
@wangpc-pp ping

**Review by wangpc-pp — APPROVED** [review]
LGTM.

**Comment by dfszabo** [comment]
We have issues with the assertion going of. Not sure, but it seems to that the machine outliner pass will create instructions without mem operands (https://github.com/llvm/llvm-project/blob/25b58c877c851bed9c34362cd69bcd8d8bb65ac4/llvm/lib/CodeGen/MachineOutliner.cpp#L957) is causing the issues. Is this possible?

**Comment by topperc** [comment]
> We have issues with the assertion going off. Not sure, but it seems to that the machine outliner pass will create instructions without mem operands
> 
> https://github.com/llvm/llvm-project/blob/25b58c877c851bed9c34362cd69bcd8d8bb65ac4/llvm/lib/CodeGen/MachineOutliner.cpp#L957
> 
> is causing the issues. Is this possible?

https://github.com/llvm/llvm-project/pull/113981

**Pull request merged**